hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommits.java

```diff
@@ -129,6 +130,52 @@ public void testThriftExceptionUnknownStateIfNotInHistoryFailureOnCommit()
         .isEqualTo(3);
   }
 
+  /**
+   * Pretends we throw a concurrent commit to allow for retry when register and overwrite an iceberg
+   * table
+   */
+  @Test
+  public void testCommitFailedExceptionOnRegisterOverwrite() {
+    Table table = catalog.loadTable(TABLE_IDENTIFIER);
+    HiveTableOperations ops = (HiveTableOperations) ((HasTableOperations) table).operations();
+
+    TableMetadata metadataV1 = ops.current();
+    String metadataV1Location = metadataV1.metadataFileLocation();
+    String metadataV1AsJson = TableMetadataParser.toJson(metadataV1);
+
+    table.updateSchema().addColumn("n", Types.IntegerType.get()).commit();
+
+    ops.refresh();
+
+    TableMetadata metadataV2 = ops.current();
+
+    assertThat(ops.current().schema().columns()).hasSize(2);
+
+    HiveTableOperations spyOps = spy(ops);
+
+    HiveCatalog spyCatalog = spy(catalog);
+    when(spyCatalog.newTableOps(TABLE_IDENTIFIER)).thenReturn(spyOps);
+    // Simulate a CommitFailedException and then allow the commit to be retried
+    doThrow(new CommitFailedException("Cannot commit: stale table metadata"))
+        .doCallRealMethod()
+        .when(spyOps)
+        .commit(any(), any());
+    spyCatalog.registerTable(TABLE_IDENTIFIER, metadataV1Location, true);
+
+    TableMetadata actual = ops.refresh();
+    assertThat(actual).as("Current metadata should have changed").isNotEqualTo(metadataV2);
+    assertThat(metadataFileExists(actual)).as("Current metadata file should still exist").isTrue();
+    assertThat(metadataFileCount(actual))
+        .as("Commit should have been successful and new metadata file should be made")
+        .isEqualTo(3);
+
+    TableMetadata expected = TableMetadataParser.fromJson(metadataV1AsJson);
+    assertThat(actual)
+        .usingRecursiveComparison()
+        .ignoringFields("metadataFileLocation")
```

Comment:
Why are we ignoring this? This must be the same

Reply:
I realized today the TableMetadataParser.to/fromJson method will skip the value from `metadataFileLocation`. So I have to ignore this when comparing the TableMetadata objects directly. However I added extra assert to ensure what you want is verified in the unit tests.